src/net_doh.c

```diff
@@ -0,0 +1,1070 @@
+    };
+    int ret = -1;
+
+    ret = nghttp2_submit_settings(session_data->session, NGHTTP2_FLAG_NONE, iv,
+                               sizeof(iv) / sizeof(*iv));
+    if (ret != 0) {
+        perf_log_warning("Could not submit https2 SETTINGS: %s", nghttp2_strerror(ret));
+        return ret;
+    }
+
+    return 0;
+}
+
+static ssize_t perf__doh_recv(struct perf_net_socket* sock, void* buf, size_t len, int flags)
+{
+    ssize_t n = 0;
+    ssize_t ret = 0;
+
+    // read TLS data here instead of nghttp2_recv_callback
+    PERF_LOCK(&self->lock);
+    if (!self->is_ready) {
+        PERF_UNLOCK(&self->lock);
+        errno = EAGAIN;
+        return -1;
+    }
+
+    n = SSL_read(self->ssl, self->recvbuf, TCP_RECV_BUF_SIZE);
+
+    if (!n) {
+        perf__doh_reconnect(sock);
+        PERF_UNLOCK(&self->lock);
+        errno = EAGAIN;
+        return -1;
+    }
+    if (n < 0) {
+        int err = SSL_get_error(self->ssl, n);
+        switch (err) {
+        case SSL_ERROR_WANT_READ:
+            errno = EAGAIN;
+            break;
+        case SSL_ERROR_SYSCALL:
+            switch (errno) {
+            case ECONNREFUSED:
+            case ECONNRESET:
+            case ENOTCONN:
+                perf__doh_reconnect(sock);
+                errno = EAGAIN;
+                break;
+            default:
+                break;
+            }
+            break;
+        default:
+            errno = EBADF;
+            break;
+        }
+        PERF_UNLOCK(&self->lock);
+        return -1;
+    }
+
+    PERF_UNLOCK(&self->lock);
+
+    // make sure we can process data
+    if (self->is_ready &&
+        self->http2 != NULL) {
+        // this will be processed by nghttp2 callbacks
+        // self->recvbuf holds the payload
+        PERF_LOCK(&self->lock);
+
+        ret = nghttp2_session_mem_recv(self->http2->session, 
+                                       (uint8_t*) self->recvbuf, 
+                                       n);
+     
+        if (ret < 0) {
+            // 
+            perf_log_warning("nghttp2_session_mem_recv failed: %s", 
+                            nghttp2_strerror((int) ret));
+
+            http2_session_free(sock);
+            PERF_UNLOCK(&self->lock);     
+            return -1;
+        }
+
+        // need to execute nghttp2_session_send if the receive ops triggered data frames 
+        ret = nghttp2_session_send(self->http2->session);
+        if (ret < 0) {
+            perf_log_warning("nghttp2_session_send failed: %s", nghttp2_strerror((int) ret));
+            PERF_UNLOCK(&self->lock);
+            return -1;
+        }
+
+        PERF_UNLOCK(&self->lock);
+    }
+    
+    PERF_LOCK(&self->lock);
+
+    if (self->http2 &&
+        self->http2->dnsmsg_completed) {
+        if (self->http2->dnsmsg_at > len) {
+            perf_log_warning("failed to process result - DNS response size");
+            PERF_UNLOCK(&self->lock);
+            return -1;
```

Comment:
Shouldn't state be reset'ed here?

If `dnsmsg_at > len` then this will probably loop infinitive.